src/main/java/io/vertx/core/eventbus/impl/Handlers.java

```diff
@@ -11,21 +11,57 @@
 
 package io.vertx.core.eventbus.impl;
 
-import java.util.List;
-import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.*;
 import java.util.concurrent.atomic.AtomicInteger;
 
 /**
  * @author <a href="http://tfox.org">Tim Fox</a>
  */
-public class Handlers {
+public class Handlers implements Iterable<HandlerHolder> {
 
-  private final AtomicInteger pos = new AtomicInteger(0);
-  public final List<HandlerHolder> list = new CopyOnWriteArrayList<>();
+  private final AtomicInteger pos;
+  private final HandlerHolder[] elements;
```

Comment:
Why not simply an ArrayList? It seems to me code below would be simplified without performance or memory penalty. Also you get the iterator for free without wrapping the array in a list any time you publish a message.

Reply:
ArrayList has mutable array field, using final we are sure the object array is safely published

Reply:
But it is no longer published. That was the purpose of making `Handlers` implement `Iterable`, wasn't it?

Reply:
that's for the stop sequence when the event bus is closed

Reply:
No, it's also used if you _publish_ (see `deliverMessageLocally`)

Reply:
I got you. Read the comments stream again 😀